plutus-core/plutus-ir/src/PlutusIR/Transform/LetFloatIn.hs

```diff
@@ -0,0 +1,188 @@
+{-# LANGUAGE BangPatterns      #-}
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TupleSections     #-}
+{-# LANGUAGE ViewPatterns      #-}
+
+{- | Float bindings inwards. A binding can be floated if it is a lazy binding, or
+ a strict binding whose RHS is a value. Doing so does not have much impact on the
+ program size since it simply moves bindings around without changing the number
+ of bindings, but it may lead to more efficient evaluation.
+
+ Unlike the float-in in GHC, we /can/ float bindings into lambda abstractions,
+ because doing so cannot duplicate work. Therefore we simply float a binding
+ inwards to the deepest possible position.
+
+ We avoid moving a strict binding whose RHS is not a value, because doing so
+ may either change the semantics (if the RHS has effects) or duplicate work
+ (if it is floated into a lambda abstraction).
+
+ Floating inwards is done in a single pass. The subterms are processed bottom-up,
```

Comment:
I actually think it might be nicer to do it in two. Annotating the AST with used variable sets is generically useful, and I think it would make the code clearer. We already do lots of passes, one more isn't so bad, it's only linearly more work.

Reply:
Thinking about it some more, I can think of at least one reason not to do this. Floating in changes the set of used variables, so if you have something like this:
```
let x = 1
in \y .
let z = x
in \p .
z
```
you want to float in both let bindings the whole way. But if you did the free variable computation first then you could only do this:
```
\y .
let x = 1
in \p .
let z = x
in z
```
because you think that the `\p . ...` uses `x`, even though it doesn't any more.

But it would be good to explain this (https://www.michaelpj.com/blog/2022/04/24/on-commenting-code.html#road-not-travelled-comments).

Reply:
Well, your code _does_ have to update the annotated sets as it pushes things in, so I'm not sure this version is simpler. Indeed, I'm not 100% confident it's all doing the right thing... if we had a pass for annotating free/used variables then we could also test that independently!

Reply:
> we could also test that independently!

True. I just doubt that having a first pass makes the second pass much simpler. During the second pass the free variables will need to be maintained when we push a binding down the tree, and that doesn't sound much easier than calculating the free variables from scratch!